File: src/test/java/org/railwaystations/rsapi/core/model/StationTest.java
package org.railwaystations.rsapi.core.model;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.assertj.core.api.Assertions.assertThat;

class StationTest {

    private static final Station.Key TEST_KEY = new Station.Key("", "0");

    @ParameterizedTest
    @CsvSource({
            "test,    false",
            ",        true"
    })
    void appliesToNullPhotographer(String photographer, boolean expectedAppliesTo) {
        var station = createStationTestFixtureBuilder()
                .build();
        assertThat(station.appliesTo(photographer)).isEqualTo(expectedAppliesTo);
    }

    @ParameterizedTest
    @CsvSource({
            "test,    true",
            ",        true"
    })
    void appliesToPhotographer(String photographer, boolean expectedAppliesTo) {
        var station = createStationTestFixtureBuilder().build();
        station.getPhotos().add(createTestPhoto());
        assertThat(station.appliesTo(photographer)).isEqualTo(expectedAppliesTo);
    }

    private Station.StationBuilder createStationTestFixtureBuilder() {
        return Station.builder()
                .key(TEST_KEY)
                .title("");
    }

    private Photo createTestPhoto() {
        return Photo.builder()
                .stationKey(TEST_KEY)
                .urlPath("URL")
                .photographer(User.builder()
                        .id(0)
                        .name("test")
                        .url("photographerUrl")
                        .license(License.CC0_10)
                        .ownPhotos(true)
                        .sendNotifications(true)
                        .build())
                .license(License.CC0_10)
                .build();
    }

}

File: core/src/test/kotlin/org/railwaystations/rsapi/core/model/StationTest.kt
package org.railwaystations.rsapi.core.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.time.Instant

private val TEST_KEY = Station.Key("", "0")

internal class StationTest {
    @ParameterizedTest
    @CsvSource(
        "test,    false", ",        true"
    )
    fun appliesToNullPhotographer(photographer: String?, expectedAppliesTo: Boolean) {
        val station = createValidStation()
        assertThat(station.appliesTo(photographer)).isEqualTo(expectedAppliesTo)
    }

    @ParameterizedTest
    @CsvSource(
        "test,    true", ",        true"
    )
    fun appliesToPhotographer(photographer: String?, expectedAppliesTo: Boolean) {
        val station = createValidStation().copy(
            photos = listOf(createTestPhoto())
        )
        assertThat(station.appliesTo(photographer)).isEqualTo(expectedAppliesTo)
    }

    private fun createValidStation(): Station {
        return Station(
            key = TEST_KEY,
            title = "",
        )
    }

    private fun createTestPhoto(): Photo {
        return Photo(
            stationKey = TEST_KEY,
            primary = true,
            urlPath = "URL",
            photographer = User(
                name = "test",
                url = "photographerUrl",
                license = License.CC0_10,
                ownPhotos = true,
            ),
            createdAt = Instant.now(),
            license = License.CC0_10,
        )
    }

}